### PR TITLE
Add lcov compatible coverage reporting to clarity-cli

### PR DESCRIPTION
**Pull request by kantai**
Starting this as a draft PR.

This PR adds an option to two `clarity-cli` commands (`launch` and `execute`)

```
-c coverage-folder-name
```

If provided, those `clarity-cli` commands will output execution traces to the given folder. This folder can be used in a new command `make_lcov` to produce an `lcov` compatible trace file with line-coverage counts.

This is implemented with pretty minimal impact on the Clarity VM. It adds a new `CoverageReporter` struct, which is an optional field in the vm's `GlobalContext`. If present, the Clarity `eval` command notifies the reporter whenever a sexp is evaluated.

## Usage example

You can see how this works by doing some testing with one of the sample contracts:

```
$ clarity-cli initialize /tmp/vm-state.db
Database created.

$ clarity-cli launch -c /tmp/clar-cov SPYZTFNY7GSMVZ211ZGNDQ5ACBY2FP6YCXV3G9MZ.tokens $PWD/sample-contracts/tokens.clar /tmp/vm-state.db 
Contract initialized!

$ clarity-cli execute -c /tmp/clar-cov/ /tmp/vm-state.db SPYZTFNY7GSMVZ211ZGNDQ5ACBY2FP6YCXV3G9MZ.tokens mint! SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR u100000
Transaction executed and committed. Returned: u100000
[]

$ clarity-cli execute -c /tmp/clar-cov/ /tmp/vm-state.db SPYZTFNY7GSMVZ211ZGNDQ5ACBY2FP6YCXV3G9MZ.tokens token-transfer SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR \'SM2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQVX8X0G u0
Aborted: "must transfer positive balance and possess funds"

$ clarity-cli execute -c /tmp/clar-cov/ /tmp/vm-state.db SPYZTFNY7GSMVZ211ZGNDQ5ACBY2FP6YCXV3G9MZ.tokens token-transfer SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR \'SM2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQVX8X0G u100
Transaction executed and committed. Returned: u100
[]

$ clarity-cli make_lcov /tmp/clar-cov/ /tmp/clarity-coverage.lcov
```

Once that `.lcov` file is produced, you can use the `genhtml` lcov tool to produce a friendly coverage output (you could probably also upload to coveralls/codecov pretty easily:

```
$ genhtml /tmp/clarity-coverage.lcov -o /tmp/coverage/
Reading data file /tmp/clarity-coverage.lcov
genhtml: WARNING: invalid characters removed from testname in tracefile /tmp/clarity-coverage.lcov
Found 1 entries.
Found common filename prefix "/Users/aaron/devel/stacks-blockchain"
Writing .css and .png files.
Generating output.
Processing file sample-contracts/tokens.clar
Writing directory view page.
Overall coverage rate:
  lines......: 94.7% (18 of 19 lines)
  functions..: no data found
```

This is what that looks like:

Screen Shot 2021-04-16 at 12 23 14 PM

### CODE REVIEW

File: src/clarity.rs [hunk 1]
```diff
@@ -489,6 +492,18 @@ fn consume_arg(
 }
 
 pub fn invoke_command(invoked_by: &str, args: &[String]) {
+    let mut args = Vec::from(args);
+
+    let coverage_folder =
+        if let Some(coverage_flag_index) = args[1..].iter().position(|arg| arg == "-c") {
```
Comment (jcnelson) [inline src/clarity.rs:498]: Do you want to use `consume_arg()` here instead? It does all the same kinds of lifting.

File: src/vm/coverage.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        contract: &QualifiedContractIdentifier,
+    ) {
+        if expr.match_list().is_some() {
+            // don't count the whole list expression: wait until we've eval'ed the
+            //   list components
+            return;
+        }
+
+        // other sexps can only span 1 line
+        let line_executed = expr.span.start_line;
```
Comment (jcnelson) [inline src/vm/coverage.rs:57]: I'm confused by this.  How would this be reported in the coverage data?

```
(+
   u1
   u2
   u3
 )
 ```
 
 Is this 5 lines, or one line?

Comment (kantai) [reply src/vm/coverage.rs:57]: 4 lines

**Review by jcnelson — APPROVED** [review]
LGTM! Thanks @kantai!

**Comment by kantai** [comment]
Tagging @lgalabru for a review!

**Review by lgalabru — APPROVED** [review]
This PR looks good to me!
Note that the approach in the REPL/Clarinet is a bit different: looking at the lcov spec, I noticed that we could be splitting the coverage report into sections, that you can label with the name of your tests, which sounded like an interesting approach for diving into coverage reports. 
Also, all the data is also collected and kept in memory, without touching the disk for intermediary artifacts.
This PR was extremely useful, thanks!

**Pull request merged**